rotkehlchen/interfaces.py

```diff
@@ -70,20 +72,73 @@ def can_query_history(
         """Checks if it's okay to query historical price"""
 
     @abc.abstractmethod
-    def query_historical_price(
+    def _query_historical_price(
             self,
             from_asset: Asset,
             to_asset: Asset,
             timestamp: Timestamp,
     ) -> Price:
-        """An oracle implements this to return a historical price from_asset to to_asset at time.
+        """Implementation specific historical price query logic.
 
         If no price can be found may raise:
         - PriceQueryUnsupportedAsset
         - NoPriceForGivenTimestamp
         - RemoteError
         """
 
+    @abc.abstractmethod
+    def oracle_source(self) -> HistoricalPriceOracle:
+        """Source enum used as key for storing and retrieving cached prices"""
+
+    @abc.abstractmethod
+    def max_price_age_in_seconds(self) -> int:
+        """Maximum time difference in seconds for which a historical price is considered valid"""
+
+    def query_historical_price(
+            self,
+            from_asset: Asset,
+            to_asset: Asset,
+            timestamp: Timestamp,
+    ) -> Price:
+        """Fetch historical price with caching.
+        Uses cached price if fresh enough, otherwise queries and caches new price.
+
+        May raise:
+        - PriceQueryUnsupportedAsset
+        - NoPriceForGivenTimestamp
+        - RemoteError
+        """
+        price_cache_entry = GlobalDBHandler.get_historical_price(
```

Comment:
Why do you need to check the global DB for the given oracle? This makes no sense.

So this is called also in the for loop for each oracle? So you would check per oracle if a price is cached? That's useless.

There should be only a single check like this. And that shoudl be at the very start of the loop or if it's not called in the oracle loop at the start. But it should not matter from which oracle the price is saved in the DB, should it?